Added a new chat message block with custom properties

joueur883 opened the pull request:
I added a new procedure block that allows sending chat messages with custom color and formatting (bold, italic, underlined) in a more intuitive way.

This simplifies the process compared to using the "Execute command" block with a `/tellraw @a` command.

<img width="842" height="86" alt="block" src="https://github.com/user-attachments/assets/fe46bde9-9f88-4815-9341-4a3cc23e64ba" />

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/procedures/world_send_chat_advanced.java.ftl:
```diff
@@ -0,0 +1,6 @@
+if (world instanceof ServerLevel _level)
```
This is a bit cluttered code, and also when adding procedure block, it is not supposed to be a mask for a command content. If you want to make a procedure block to replace command, it should directly call underlying systems that a command invocation calls, to avoid the overhead of command parsing and while command system the current approach does.

KlemenDEV reviewed (changes requested):
Comment applies to both versions

joueur883 commented:
I've refactored the procedure block to use Minecraft's internal APIs instead of command parsing. The message now uses `Component.literal()` with `withStyle()` for color and formatting (bold, italic, underline), and is broadcasted via `broadcastSystemMessage()`. Color mapping is handled via FreeMarker if/elseif due to template limitations. Let me know if you'd prefer a different structure or if further adjustments are needed!

KlemenDEV commented on plugins/mcreator-core/procedures/world_send_chat_advanced.json:
```diff
@@ -11,14 +11,14 @@
       "options": [
         ["white", "white"],
```
you can already map to rgb values here in FTL to avoid the chain of if statements

joueur883 commented:
I’ve mapped the RGB values and merged the advanced block into the existing world_send_chat block for cleaner integration.
Let me know if you need any further tweaks!

KlemenDEV commented:
This is not what I meant. JSON definition of procedure block (.json file) defines human name and value. Just use the value of rgb value directly there, no mapping in FTL file

joueur883 commented:
I've updated the JSON to use RGB values directly as you suggested, and removed the mapping logic from the FTL file. Let me know if this version aligns better with what you had in mind!

joueur883 commented:
I see the problem! I forgot to add default values to the fields for retro-compatibility. I'm fixing it right now and will push the update in a few minutes.

joueur883 commented:
I think editing the original block, even with default values breaks retro-compatibility because those fields didn’t exist in previous versions.

KlemenDEV commented:
> I think editing the original block, even with default values breaks retro-compatibility because those fields didn’t exist in previous versions.

Many times in the past we added new fields to existing procedure blocks and converter is not needed for this. You just need to provide all default values correctly. What do the test logs say?

> and everything builds fine locally. I believe it's ready now!

Did you run tests locally? Please run unit tests locally too, this way you can better understand what is going on

KlemenDEV commented:
If you check test logs, they tell quite explicitly what the problem is:

<img width="793" height="96" alt="image" src="https://github.com/user-attachments/assets/75b8255f-0efa-4faa-89b7-970240df6dff" />

KlemenDEV commented:
> Also I ran local tests and saw 12 build failures, but none of them seem related to this block. From what I can tell, the problem is resolved.

Test should not fail to build and should pass

joueur883 commented on plugins/generator-1.21.1/neoforge-1.21.1/procedures/world_send_chat.java.ftl:
```diff
@@ -1,2 +1,15 @@
+if (world instanceof ServerLevel _level) {
+	MutableComponent message = Component.literal(${input$text!"Message"});
+
+	<#if field$color?? && field$bold?? && field$italic?? && field$underlined??>
```
I initially added it as a safety check to ensure everything would work properly, but you're right, it's probably redundant now. I can remove it!

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/procedures/world_send_chat.java.ftl:
```diff
@@ -1,2 +1,13 @@
-	world.getServer().getPlayerList().broadcastSystemMessage(Component.literal(${input$text}), false);
\ No newline at end of file
+if (world instanceof ServerLevel _level) {
+	MutableComponent message = Component.literal(${input$text!"Message"});
```
Since this all code is part of the same system, you should leave out new line spacing here and below too, applies to both versions

KlemenDEV commented on plugins/mcreator-core/procedures/world_send_chat.json:
```diff
@@ -5,24 +5,67 @@
       "check": "String"
     },
+    {
+      "type": "field_dropdown",
```
Since you can technically define any color, I have added new color chooser field to the core: https://github.com/MCreator/MCreator/commit/52bbb9380314679d4e633e9f15f15a7dfa8cdc6d

Please adapt this block to use it and merge master branch in your PR so it will work in your PR

KlemenDEV reviewed (changes requested):
Two comments

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/procedures/world_send_chat.java.ftl:
```diff
@@ -1,2 +1,12 @@
\ No newline at end of file
+if (world instanceof ServerLevel _level) {
+	MutableComponent message = Component.literal(${input$text!"Message"});
+	<#assign newcolor = field$color?substring(1)>
```
Avoid assign, just do it directly, since assign carries some overhead

KlemenDEV replied:
You also still need to specify default value. Did you run the tests?

joueur883 commented:
I now understand how important it is to run local tests before marking a PR as ready. I’ll fix the issue and make sure everything is properly tested locally before marking it ready again.

joueur883 commented:
Is this log normal?
<img width="1558" height="55" alt="screenshot" src="https://github.com/user-attachments/assets/0e386fb3-8969-46bf-88b2-320001aacff7" />

KlemenDEV commented:
PR tested (also with pre-PR workspaces) and is ready for review?

Spectrall368 reviewed (approved):
Everything works,I found 2 more methods that could be added to the pr if you want:
`withStrikethrough` and `withObfuscated`

KlemenDEV commented on plugins/mcreator-core/procedures/world_send_chat.json:
```diff
@@ -5,24 +5,54 @@
     "toolbox_init": [
-      "<value name=\"text\"><block type=\"text\"><field name=\"TEXT\">Message</field></block></value>"
+      "<value name=\"text\"><block type=\"text\"><field name=\"TEXT\">Message</field></block></value>",
+      "<value name=\"color\"><block type=\"field_color\"></block></value>",
```
Looking at other procedure blocks, we don't really populate fields. They are defined including their default value already in their definitions. Omit fields from here

KlemenDEV commented on plugins/generator-1.21.8/neoforge-1.21.8/procedures/world_send_chat.java.ftl:
```diff
@@ -1,2 +1,11 @@
\ No newline at end of file
+if (world instanceof ServerLevel _level) {
+	MutableComponent message = Component.literal(${input$text!"Message"});
+	message.withStyle(style -> style
```
It may make sense to only call e.g. withBold if bold is enabled, so we don't do reduntant calls withBold(false) and similar for false values. Applies to all parameters and both versions. You could even do a direct call like before if all 3 are unchecked and color is #ffffff

KlemenDEV commented on plugins/mcreator-core/procedures/world_send_chat.json:
```diff
@@ -5,24 +5,50 @@
     "toolbox_id": "worldmanagement",
     "toolbox_init": [
-      "<value name=\"text\"><block type=\"text\"><field name=\"TEXT\">Message</field></block></value>"
+      "<value name=\"text\"><block type=\"text\"><field name=\"TEXT\">Message</field></block></value>",
```
Unnnecessary comma

joueur883 replied:
Yes I saw it but I forgot to remove it when I committed so I will fix it now

KlemenDEV reviewed (approved):
LGTM, thanks!
